### PR TITLE
Restructure for the 1.0 Call API

### PR DESCRIPTION
**Pull request by levb**
Refactored for https://github.com/mattermost/mattermost-plugin-apps/pull/296

TODO
- [ ] OpenFaas support
- [ ] Try using `Trouble Ticket` API instead of `Tables`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,56 +1,277 @@
-BASE?=http://localhost:3000
-ADDR?=:3000
-
-GO ?= go
+GO ?= $(shell command -v go 2> /dev/null)
```
Comment (levb) [inline Makefile:1]: updated to that of a plugin

File: build/custom.mk [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Include custom targets and environment variables here
```
Comment (levb) [inline build/custom.mk:1]: custom build targets for the app: `run` for debugging as http, `dist` from the makefile is used for the plugin build.

File: function/app.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+func Init(mode string, r *mux.Router, log utils.Logger) {
+	app := App{
+		mode: mode,
+		App: *goapp.NewApp(r, log).
+			WithManifest(root.AppManifest).
+			WithStatic(root.Static).
+			WithIcon(root.AppManifest.Icon),
+	}
+
+	app.HandleCall("/install", install)
```
Comment (levb) [inline function/app.go:40]: No-op for now, need one?

File: function/servicenow.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		return "", err
+	}
+
+	resp, err := c.client.Post(apiURL(creq, "/api/now/table/"+table), "application/json", bytes.NewReader(b))
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	tok, _ := c.tokenSource.Token()
```
Comment (levb) [inline function/servicenow.go:76]: This (refreshing the cached access token) needs to be generalized somehow, no idea yet on what's the best way to package.

File: goapp/app.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
+package goapp
```
Comment (levb) [inline goapp/app.go:1]: This should mature a little and move to `mattermost-plugin-apps/apps/goapp` perhaps?

File: manifest-v0.7.0.json [hunk 6]
```diff
@@ -0,0 +1,49 @@
+{
```
Comment (levb) [inline manifest-v0.7.0.json:1]: Used for Mattermost Cloud distro

**Comment by levb** [comment]
@hanzei @larkox any idea what did I mess up vis-a-vis the build pipeline and how to fix it? Should I log a separate ticket to add CI support to build the `plugin`, `aws`, and `cloud` bundles? Help would be much appreciated.

File: build/custom.mk [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+.PHONY: dist-aws
+dist-aws:
+	rm -rf dist/aws && mkdir -p dist/aws
+	cd aws ; \
+		CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build $(GO_BUILD_FLAGS) -o ../dist/aws/servicenow .
+	cp manifest.json dist/aws
+	cp -r static dist/aws
+	cd dist/aws ; \
+		zip -m servicenow.zip servicenow ; \
+		zip -rm ../$(AWS_BUNDLE_NAME) manifest.json static servicenow.zip
```
Comment (hanzei) [inline build/custom.mk:31]: I think CI expects the output to be named `bundle.zip`

File: .circleci/config.yml [hunk 8]
```diff
@@ -1,65 +1,40 @@
...
       - persist_to_workspace:
           root: dist
           paths:
-            - "bundle.zip"
+            - "*.zip"
+            - '*.tar.gz'
+            - release-notes.md
       - store_artifacts:
-          path: dist/bundle.zip
+          path: dist
```
Comment (hanzei) [inline .circleci/config.yml:37]: What else is in `dist`?

**Review by hanzei — APPROVED** [review]
Nice work :+1:

File: .circleci/config.yml (hunk 8 above)
Comment (levb) [reply .circleci/config.yml:37]: image

**Comment by levb** [comment]
@mickmister @catalintomai @larkox a gentle ping for a second review?

**Review by catalintomai — APPROVED** [review]

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Confirmed the app can be deployed and is functional 
- Successfully Connected a user and confirmed  ticket creation is working
- User disconnect and reconnect is working as expected

LGTM!

**Review by mickmister — COMMENTED** [review]
LGTM

**Pull request merged**